impl/src/main/java/org/jboss/weld/xml/BeansXmlValidator.java

```diff
@@ -0,0 +1,158 @@
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.xml.XMLConstants;
+import javax.xml.transform.Source;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import javax.xml.validation.Validator;
+
+import org.jboss.weld.logging.XmlLogger;
+import org.xml.sax.ErrorHandler;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+
+/**
+ * This validator is thread safe and suitable for sharing between threads.
+ *
+ * @author Martin Kouba
+ */
+public class BeansXmlValidator implements ErrorHandler {
+
+    private static final Source[] EMPTY_SOURCE_ARRAY = new Source[0];
+    private static final String ROOT_ELEMENT_NAME = "beans";
+    // See also http://www.w3.org/TR/2001/REC-xmlschema-1-20010502/#cvc-elt
+    private static final String VALIDATION_ERROR_CODE_CVC_ELT_1 = "cvc-elt.1";
+
+    private final Schema cdi11Schema;
+
+    private final Schema cdi20Schema;
+
+    public BeansXmlValidator() {
+        SchemaFactory factory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+        cdi11Schema = initSchema(factory, XmlSchema.CDI11_SCHEMAS);
+        cdi20Schema = initSchema(factory, XmlSchema.CDI20_SCHEMAS);
+    }
+
+    public void validate(URL beansXml, ErrorHandler errorHandler) {
+        if (beansXml == null) {
+            throw XmlLogger.LOG.loadError("unknown", null);
+        }
+        if (errorHandler == null) {
+            errorHandler = this;
+        }
+        Schema schema = cdi20Schema;
+
+        // First quick check of beans.xml to find out version
+        try (InputStream in = beansXml.openStream()) {
+            if (in.available() == 0) {
+                // The file is just acting as a marker file
+                return;
+            }
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(in, Charset.forName(StandardCharsets.UTF_8.name())))) {
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    if (line.contains(XmlSchema.CDI11.getFileName())) {
+                        schema = cdi11Schema;
+                        break;
+                    } else if (line.contains(XmlSchema.CDI20.getFileName())) {
+                        break;
+                    }
+                }
+            }
+        } catch (IOException e) {
+            throw XmlLogger.LOG.loadError(beansXml, e);
+        }
+
+        if (schema == null) {
+            return;
+        }
+        Validator validator = schema.newValidator();
+        validator.setErrorHandler(errorHandler);
+        try {
+            validator.validate(new StreamSource(beansXml.toExternalForm()));
```

Comment:
If you wanted to be able to close the stream here, maybe do:

    try (final InputStream in = new BufferedInputStream(beansXml.openStream()) {
        validator.validate(in);
    } catch (SAXException | IOException e) {
        // No-op - validation is optional
    }

Reply:
…unless something magic™ happens 😄 by converting a URL to string form when you supply it to the `StreamSource` constructor?  If you already have the URL, I would think just open its stream again here.

Reply:
Hm, `javax.xml.validation.Validator` only accepts `javax.xml.transform.Source` so something like this looks ok:
````
try (final InputStream in = new BufferedInputStream(beansXml.openStream()) {
    validator.validate(new StreamSource(in));
} catch (SAXException | IOException e) {
    // No-op - validation is optional
}
````